File: Python/백준/Bronze/2798. 블랙잭/블랙잭.py
import sys
input = sys.stdin.readline

def solve():
    N, M = map(int, input().split())
    cards = list(map(int, input().split()))
    best = 0
    # 세 장을 고르는 모든 조합 탐색
    for i in range(N - 2):
        for j in range(i + 1, N - 1):
            for k in range(j + 1, N):
                s = cards[i] + cards[j] + cards[k]
                if s <= M and s > best:
                    best = s
                # 만약 정확히 M이면 더 이상 클 수 없으므로 조기 종료 가능
                if best == M:
                    print(M)
                    return
    print(best)

if __name__ == "__main__":
    solve()

File: C99/백준/Bronze/2798. 블랙잭/블랙잭.c
#include <stdio.h>

int main() {
    int N, M;
    scanf("%d %d", &N, &M);
    int cards[101];
    for (int i = 0; i < N; i++) {
        scanf("%d", &cards[i]);
    }
    int best = 0;
    for (int i = 0; i < N - 2; i++) {
        for (int j = i + 1; j < N - 1; j++) {
            for (int k = j + 1; k < N; k++) {
                int sum = cards[i] + cards[j] + cards[k];
                if (sum <= M && sum > best) {
                    best = sum;
                }
                if (best == M) {
                    printf("%d\n", M);
                    return 0;
                }
            }
        }
    }
    printf("%d\n", best);
    return 0;
}